### PR TITLE
Add widget -> riot postMessage API

### PR DESCRIPTION
**Pull request by rxl881**

### CODE REVIEW

**Comment by rxl881** [comment]
Depends upon https://github.com/matrix-org/modular-widgets/pull/26

File: src/WidgetMessaging.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+let listenerCount = 0;
+let messageEndpoints = [];
+
+/**
+ * Handle widget postMessage events
+ * @param  {Event} event Event to handle
+ * @return {undefined}
+ */
+function onMessage(event) {
```
Comment (turt2live) [inline src/WidgetMessaging.js]: This should probably always return something back to the widget, like the scalar API. 

Ie:
```
{ success: true, .. other params .. }
// other params would be the original request
```

```
{ success: false, error: "Unknown action", .. other params .. }
```

**Comment by turt2live** [comment]
General request: Some kind of arbitrary version endpoint so that widgets may react to future endpoints or changes in the API. The Scalar API currently lacks this, and it's quite annoying.

here's what I'd imagine:

request:
```
{
  action: "get_widget_api_versions"
}
```

response:
```
{
  success: true,
  action: "get_widget_api_versions",
  supported_versions: ["1.0.0"]
}
```

`supported_versions` is intentionally an array so that breaking changes can be made without breaking every widget imaginable. The version should be semantic.

File: src/WidgetMessaging.js (hunk 1 above)
Comment (rxl881) [reply src/WidgetMessaging.js]: Thanks for pointing this out @turt2live - I've added responses (and comments documenting the API) to bring it in line with the scalarMessaging API (slightly different structure to the one that you describe above).

**Comment by rxl881** [comment]
@turt2live - Thanks for your request regarding supported API versions. I'd love to see us generally doing more of this. I've added two additional endpoints, one to get the current API version, and one to get the supported versions (largely) as suggested.

**Comment by turt2live** [comment]
Much appreciated! 

and +1 to more future proofing.

File: src/WidgetMessaging.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    if (!event.origin) { // Handle chrome
+        event.origin = event.originalEvent.origin;
+    }
+
+    // Event origin is empty string if undefined
+    if (
+        event.origin.length === 0 ||
+        trustedEndpoint(event.origin) ||
```
[the remark this answers is not in the document]
Comment (rxl881) [reply src/WidgetMessaging.js]: Good catch! Fixed.

File: src/WidgetMessaging.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        if (messageEndpoints.filter(function(endpoint) {
+            if (endpoint.endpointUrl == origin) {
+                return true;
+            }
+        }).length > 0) {
+            return true;
+        }
+    }
+
+    return false;
```
[the remark this answers is not in the document]
Comment (rxl881) [reply src/WidgetMessaging.js]: Nice. I didn't know about .some() 👍 .

Yup, on the strict equality checking (need to erase that muscle memory ;) ). I've corrected the other slips.

File: src/WidgetMessaging.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import dis from './dispatcher';
+
+let listenerCount = 0;
+let messageEndpoints = [];
```
[the remark this answers is not in the document]
Comment (rxl881) [reply src/WidgetMessaging.js]: Good point. However, this is exactly the same pattern as from the existing matrix-react-sdk ScalarMessaging class, and I can't see an easy / obvious way around this. Can you suggest a workaround, or are you happy to leave this as is for the time being (it isn't a problem for the existing use cases)?

[the remark this answers is not in the document]
Comment (rxl881) [reply src/WidgetMessaging.js]: Right, as discussed IRL, that should be sorted now. Thanks for the tip @dbkr

File: src/WidgetMessaging.js [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+ * @return {boolean} True if endpoint was successfully removed
+ */
+function removeEndpoint(widgetId, endpointUrl) {
+    const u = URL.parse(endpointUrl);
+    if (!u || !u.protocol || !u.host) {
+        console.warn("Invalid origin");
+        return;
+    }
+
+    console.warn("Origin url:", u);
```
Comment (dbkr) [inline src/WidgetMessaging.js]: was this meant to be left in?

**Comment by rxl881** [comment]
Hopefully all the feedback is addressed now. @dbkr, can you please take another look?

File: src/WidgetMessaging.js (hunk 5 above)
Comment (rxl881) [reply src/WidgetMessaging.js]: Looks like I forgot to push the change cleaning that up. Pushed now. Thanks!

**Pull request merged**